src/tpm/tpm_wrapper.rs

```diff
@@ -0,0 +1,271 @@
+use std::ffi::CString;
+use std::mem::MaybeUninit;
+use std::os::raw::c_char;
+use std::ptr::null_mut;
+use std::sync::Mutex;
+use libc::{c_void};
+
+use crate::{error, tpm};
+use tpm::Error as TpmError;
+
+use tss2::{
+    FAPI_CONTEXT, TSS2_RC, TSS2_RC_SUCCESS, ESYS_TR, FAPI_ESYSBLOB_CONTEXTLOAD, TPMS_CONTEXT,
+    TSS2_TCTI_CONTEXT, ESYS_CONTEXT, TPM2B_ECC_POINT, ESYS_TR_PASSWORD, ESYS_TR_NONE,
+    TPM2B_PUBLIC, TPM2B_NAME,
+    UINT16,
+    Tss2_MU_TPMS_CONTEXT_Unmarshal,
+    Esys_Initialize, Esys_ContextLoad, Esys_ECDH_ZGen, Esys_FlushContext, Esys_Finalize, Esys_ReadPublic,
+    Fapi_Initialize, Fapi_Finalize, Fapi_GetEsysBlob, Fapi_GetTcti, Fapi_Sign
+};
+
+pub type Result<T = ()> = std::result::Result<T, error::Error>;
+
+static mut TPM_CTX: Option<Mutex<*mut FAPI_CONTEXT>> = None;
+
+fn tpm<'a>() -> &'a Mutex<*mut FAPI_CONTEXT> {
+    unsafe { TPM_CTX.as_ref().unwrap() }
+}
+
+fn with_tpm<F, R>(f: F) -> R
+    where
+        F: FnOnce(*mut FAPI_CONTEXT) -> R,
+{
+    let tpm_ctx = tpm().lock().unwrap();
+    f(*tpm_ctx)
+}
+
+pub fn tpm_init() -> Result {
+    unsafe {
+        let mut tpm_ctx: *mut FAPI_CONTEXT = MaybeUninit::uninit().assume_init();
+        let res = Fapi_Initialize(&mut tpm_ctx as *mut *mut FAPI_CONTEXT, null_mut());
+        if res != TSS2_RC_SUCCESS {
+            return Err(TpmError::tpm_error(String::from("Fapi_Initialize"), res));
+        }
+
+        TPM_CTX = Some(Mutex::new(tpm_ctx));
+    }
+
+    Ok(())
+}
+
+pub fn tpm_deinit() {
+    unsafe {
+        let mut tpm_ctx = *tpm().lock().unwrap();
+        Fapi_Finalize(&mut tpm_ctx);
+        TPM_CTX = None;
+    }
+}
+
+fn free_esys_resources (esys_ctx: &mut *mut ESYS_CONTEXT, esys_key_handle: ESYS_TR) -> Result {
+    unsafe {
+        if esys_key_handle != u32::MAX {
+            let res = Esys_FlushContext(*esys_ctx, esys_key_handle);
+            if res != TSS2_RC_SUCCESS {
+                return Err(TpmError::tpm_error(String::from("Esys_FlushContext"), res));
+            }
+        }
+
+        Esys_Finalize(&mut *esys_ctx);
+    }
+
+    Ok(())
+}
+
+pub fn public_key(key_path: &String) -> Result<Vec<u8>> {
+    unsafe {
+        let tpm_ctx = tpm().lock().unwrap();
+        let mut esys_key_handle: ESYS_TR = u32::MAX;
+        let mut blob_type: u8 = 0;
+        let mut esys_blob: *mut u8 = null_mut();
+        let mut blob_sz: tss2::size_t = 0;
+        let mut offset: tss2::size_t = 0;
+        let mut result = Fapi_GetEsysBlob(*tpm_ctx,
+                                 CString::new(key_path.as_bytes()).unwrap().as_ptr(),
```

Comment:
If you're going to unwrap the result of creating a CString, might as well have the function take `&str` and do this instead?
```suggestion
                                 key_path.as_bytes().as_ptr() as *const c_char,
```